app/models/address.rb

```diff
@@ -30,18 +31,46 @@ def full_address
 
   alias to_s full_address
 
+  def geocode_with_ward
+    # TODO? Is there any point using the Geocoder gem rather than writing
+    # our own code to query postcodes.io given that using postcodes.io
+    # now defines how we use Geocoder?
+    # - Passing anything other than a postcode will cause Geocoder.search to
+    #   return an empty array.
+    # - We are accessing admin_ward directly through through
+    #   Geocoder::Result::PostcodesIo#data hash.
+
+    geo = Geocoder.search(postcode).first&.data
+
+    if geo
+      # TODO? This works. Is there a more correct (railsy) way of doing this
+      # e.g. using write_attribute ? What is the trade off?
+      self.longitude= geo['longitude']
+      self.latitude= geo['latitude']
+      self.admin_ward= geo['admin_ward']
+    end
+  end
+
   class << self
     # location - The raw location field
-    # components - Array containing parts of an event's location field
+    # components - Array containing parts of an event's location field, excluding the postcode.
     def search(location, components, postcode)
+
+      # Find the first Address whose first address line contains any one of the
+      # address lines in the components argument.
       @address = Address.where(street_address: components).first
 
-      if @address.blank? # try using coordinates to match address
-        coordinates = Geocoder.coordinates(location)
-        @address ||= Address.where(latitude: coordinates[0], longitude: coordinates[1]).first
-      end
+      # We were looking for an exact match of geocoding coordinates, but we are
+      # now using postcodes.io exclusively so a postcode match is now
+      # equivalent to a coordinate match.
+      # if @address.blank?
+      #   coordinates = Geocoder.coordinates(postcode)
+      #   @address ||= Address.where(latitude: coordinates[0], longitude: coordinates[1]).first
+      # end
 
       # Search by postcode if it is minimum length of a full postal code
+      # TODO? Are postcodes guaranteed to be in the same format? E.g A1 2BC
```

Comment:
I am not sure, we prob need a unit test for this?

Reply:
Could this be a validation step? Could we validate that all postcodes are stored in the same format so that comparisons like this are always a know quantity?

Reply:
Yep that would be ideal. Only q would be is there a method to know where the break is for outputting in the frontend? Postcodes do look weird with no space.

Reply:
According to the Market Research Society, all incodes (second part) are NAA, only the outcodes change in number of digits. https://www.mrs.org.uk/pdf/postcodeformat.pdf

Reply:
Pefecto then. Let's store space-free and add a presenter function to separate it out again.

Reply:
Also, we can choose whatever format we like for storing to DB, so we can just use the preferred front-end format. We don't lose anything. (Except paying one byte per stored postcode that we could have saved :smile: ).

Reply:
OK, will leave up to you, as long as its documented :)